src/ATen/nn_parse.py

```diff
@@ -1,49 +1,320 @@
+    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
+
+
+def get_thnn_args(thnn_function, params):
+    params_by_name = {p['name']: p for p in params}
+    dimensionality = get_dimensionality(thnn_function.name)
+
+    def arg_expr(prefix, suffix):
+        # e.g kW, kH
+        name = ARGUMENT_MAPPINGS[prefix]
+        if name not in params_by_name:
+            raise RuntimeError('missing arg "{}" in {}'.format(name, thnn_function.name))
+        param = params_by_name[name]
+        if param['type'] == 'IntList' and 'size' in param:
+            name = name + '_'
+        index = dimensionality + DIMENSION_OFFSET[suffix]
+        expr = '{}[{}]'.format(name, index)
+        return {'type': 'EXPRESSION', 'name': expr}
+
+    thnn_args = []
+    for arg in thnn_function.arguments:
+        name = arg.name
+        if name == 'state':
+            continue
+        aten_name = camel_to_snake(SUBSTITUTIONS.get(name, name))
+        if aten_name in params_by_name:
+            param = params_by_name[aten_name]
+            if arg.is_optional:
+                param['is_nullable'] = True
+            thnn_args.append(copy.deepcopy(param))
+        elif name[-1] in DIMENSION_OFFSET and name[:-1] in ARGUMENT_MAPPINGS:
+            # e.g kW, kH
+            thnn_args.append(arg_expr(name[:-1], name[-1]))
+        elif name == 'owidth' or name == 'oheight':
+            thnn_args.append(arg_expr(name[0], name[1:]))
+        elif name == 'scale':
+            thnn_args.append({'type': 'EXPRESSION', 'name': '1'})
+        else:
+            raise RuntimeError("{}: can't find binding for '{}'"
+                               .format(thnn_function.name, name))
+    return thnn_args
+
+
+def remove_unused_args(args, thnn_args):
+    def clean_name(name):
+        name = name[:name.index('[')] if '[' in name else name
+        if name.endswith('_'):
+            name = name[:-1]
+        return name
+    uses = set([clean_name(arg['name']) for arg in thnn_args])
+    uses.add('output_mask')
+    args = [arg for arg in args if arg['name'] in uses]
+    for arg in args:
+        if 'default' in arg:
+            del arg['default']
+    return args
+
+
+def unique_args(argslist):
+    result = []
+    seen = set()
+    for args in argslist:
+        for arg in args:
+            if arg['name'] in seen:
+                continue
+            seen.add(arg['name'])
+            result.append(arg)
     return result
 
 
-def function_to_declaration(func, backend):
+def function_info(name, arguments, cimpls, buffers, backends):
     return {
         'mode': 'NN',
-        'name': func.name,
+        'name': name,
         'types': ['Float', 'Double'],
-        # skip state argument...
-        'arguments': [argument_to_declaration(a) for a in func.arguments[1:]],
-        'backends': [backend],
+        'arguments': arguments,
+        'return': get_return(arguments),
+        'buffers': buffers,
+        'backends': backends,
+        'cimpls': cimpls,
```

Comment:
Can you add a comment about what cimpls is? By allowing a single function to make multiple calls to THNN stuff, it changes the behavior of the wrapper generated quite a bit and could use an explanation somehwhere.

Reply:
I'll add a comment. It contains:

`cname`: underlying function name
`arguments` actual arguments to functional call
`condition`: (optional) guard around call

The change to `function_wrapper.py` to support cimpls is relatively small:
https://github.com/zdevito/ATen/pull/83/files#diff-4ffbf4f149ccf008ff5e7e9215f1a218R741

Reply:
Yep, that sounds good. It is what I thought it was, but it took me a while to figure out that nn_parse.py populated it. Sometimes pythons dynamic typing makes things hard to understand :)